src/main/preload/externalAPI.ts

```diff
@@ -0,0 +1,459 @@
+        if (legacyEnabled) {
+            legacyOff();
+        }
+
+        return ipcRenderer.invoke(GET_APP_INFO);
+    },
+    reactAppInitialized: () => ipcRenderer.send(REACT_APP_INITIALIZED),
+
+    // Session
+    setSessionExpired: (isExpired) => ipcRenderer.send(SESSION_EXPIRED, isExpired),
+    onUserActivityUpdate: (listener) => ipcRenderer.on(USER_ACTIVITY_UPDATE, (_, userIsActive, idleTime, isSystemEvent) => listener(userIsActive, idleTime, isSystemEvent)),
+
+    // Unreads/mentions/notifications
+    sendNotification: (title, body, channelId, teamId, url, silent, soundName) =>
+        ipcRenderer.send(NOTIFY_MENTION, title, body, channelId, teamId, url, silent, soundName),
+    onNotificationClicked: (listener) => ipcRenderer.on(NOTIFICATION_CLICKED, (_, channelId, teamId, url) => listener(channelId, teamId, url)),
+    updateUnread: (isUnread) => ipcRenderer.send(UNREAD_RESULT, isUnread),
+    updateMentions: (mentionCount) => ipcRenderer.send(MENTIONS_RESULT, mentionCount),
+
+    // Navigation
+    requestBrowserHistoryStatus: () => ipcRenderer.invoke(REQUEST_BROWSER_HISTORY_STATUS),
+    onBrowserHistoryStatusUpdated: (listener) => ipcRenderer.on(BROWSER_HISTORY_STATUS_UPDATED, (_, canGoBack, canGoForward) => listener(canGoBack, canGoForward)),
+    onBrowserHistoryPush: (listener) => ipcRenderer.on(BROWSER_HISTORY_PUSH, (_, path) => listener(path)),
+    sendBrowserHistoryPush: (path) => ipcRenderer.send(BROWSER_HISTORY_PUSH, path),
+
+    // Calls widget
+    openLinkFromCallsWidget: (url) => ipcRenderer.send(CALLS_LINK_CLICK, url),
+    openScreenShareModal: () => ipcRenderer.send(DESKTOP_SOURCES_MODAL_REQUEST),
+    onScreenShared: (listener) => ipcRenderer.on(CALLS_WIDGET_SHARE_SCREEN, (_, sourceID, withAudio) => listener(sourceID, withAudio)),
+    callsWidgetConnected: (callID) => ipcRenderer.send(CALLS_JOINED_CALL, callID),
+    onJoinCallRequest: (listener) => ipcRenderer.on(CALLS_JOIN_REQUEST, (_, callID) => listener(callID)),
+    resizeCallsWidget: (width, height) => ipcRenderer.send(CALLS_WIDGET_RESIZE, width, height),
+    focusPopout: () => ipcRenderer.send(CALLS_POPOUT_FOCUS),
+    leaveCall: () => ipcRenderer.send(CALLS_LEAVE_CALL),
+    sendCallsError: (error) => ipcRenderer.send(CALLS_ERROR, error),
+
+    // Calls plugin
+    getDesktopSources: (opts) => ipcRenderer.invoke(GET_DESKTOP_SOURCES, opts),
+    onOpenScreenShareModal: (listener) => ipcRenderer.on(DESKTOP_SOURCES_MODAL_REQUEST, () => listener()),
+    shareScreen: (sourceID, withAudio) => ipcRenderer.send(CALLS_WIDGET_SHARE_SCREEN, sourceID, withAudio),
+    joinCall: (opts) => ipcRenderer.invoke(CALLS_JOIN_CALL, opts),
+    sendJoinCallRequest: (callId) => ipcRenderer.send(CALLS_JOIN_REQUEST, callId),
+    onCallsError: (listener) => ipcRenderer.on(CALLS_ERROR, (_, err, callID, errMsg) => listener(err, callID, errMsg)),
+
+    // Utility
+    unregister: (channel) => ipcRenderer.removeAllListeners(channel),
+} as DesktopAPI);
```

Comment:
It's there because `contextBridge` accepts an `any` object and when passing that object in it doesn't know that it's a `DesktopAPI` object. Doing `as DesktopAPI` actually does help catch the type errors (the whole thing will error out if you're missing something/something is ill-defined) but I can change it to the above if you think it's cleaner. 0/5.

Reply:
I'm pretty sure `as X` masks some errors because we've got plenty of cases in the web app code where tests use incomplete objects and then slap `as X` on them to resolve the typechecking errors. And then those cause runtime errors when the code changes. I think my suggestion should prevent that.

There must be some fancy type system phrase that describes what type assertions allow you to get away with and what they don't since I know they let you fake types sometimes, but other times they force you to write `as unknown as X` to do that